UNITTESTS/drivers/PwmOut/test_pwmout.cpp

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright (c) 2019 Arm Limited and affiliates.
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "gtest/gtest.h"
+#include "gmock/gmock.h"
+#include "drivers/PwmOut.h"
+
+using namespace mbed;
+
+class MbedPowerMgmtInterface {
+public:
+    virtual ~MbedPowerMgmtInterface() {}
+    virtual void sleep_manager_unlock_deep_sleep_internal() = 0;
+    virtual void sleep_manager_lock_deep_sleep_internal() = 0;
+};
+
+
+class MockMbedPowerMgmt : public MbedPowerMgmtInterface {
+public:
+    virtual ~MockMbedPowerMgmt() {}
+    MOCK_METHOD0(sleep_manager_unlock_deep_sleep_internal, void());
+    MOCK_METHOD0(sleep_manager_lock_deep_sleep_internal, void());
+
+    static MockMbedPowerMgmt *get_instance(bool create_instance)
+    {
+        if (instance == nullptr && create_instance) {
+            instance = new MockMbedPowerMgmt();
+        }
+
+        if (create_instance) {
+            ++ref_counter;
+        }
+
+        return instance;
+    }
+
+    static void release_instance()
+    {
+        --ref_counter;
+
+        if ((instance != nullptr) && (ref_counter == 0)) {
+            delete instance;
+            instance = nullptr;
+        }
+    }
+
+    MockMbedPowerMgmt(MockMbedPowerMgmt const&) = delete;
+    void operator=(MockMbedPowerMgmt const&) = delete;
+private:
+    MockMbedPowerMgmt() {}
+
+    static MockMbedPowerMgmt *instance;
+    static int ref_counter;
+};
+MockMbedPowerMgmt *MockMbedPowerMgmt::instance = nullptr;
+int MockMbedPowerMgmt::ref_counter = 0;
+
+
+void sleep_manager_unlock_deep_sleep(void)
+{
+    // Get the singleton if a test case has created one.
+    // We do not want to increment the call count unless it's done by a test
+    // case.
+    MockMbedPowerMgmt *mbed_mgmt_mock = MockMbedPowerMgmt::get_instance(false);
+    if (mbed_mgmt_mock != nullptr) {
+        // Call only if the singleton was created by a test case.
+        mbed_mgmt_mock->sleep_manager_unlock_deep_sleep_internal();
+    }
+}
+
+
+void sleep_manager_lock_deep_sleep(void)
+{
+    // Get the singleton if a test case has created one.
+    // We do not want to increment the call count unless it's done by a test
+    // case.
+    MockMbedPowerMgmt *mbed_mgmt_mock = MockMbedPowerMgmt::get_instance(false);
+    if (mbed_mgmt_mock != nullptr) {
+        // Call only if the singleton  was created by a test case.
```

Comment:
There is a lot of added complexity which does not seem necessary. Why would this function be called outside of a test case? 
`MockMbedPowerMgmt::get_instance` should take no argument and always return a pointer to the static instance so you could do `MockMbedPowerMgmt::get_instance()->sleep_manager_lock_deep_sleep_internal();`

Reply:
It will be far more constructive to ask why something was done a certain way in order to understand the need and potentially suggest a simpler way (if one exists) instead of saying that it is unnecessary or how something should or should not be.

Why would the function be called outside of a test case? Because the `PwmOut` object is instantiated in `Setup` and the constructor calls `sleep_manager_lock_deep_sleep`.

Why is it created in `Setup`? So the expectation count in the test cases does not include the call by the constructor. The test cases are for unit tests, to check the call counts by the `resume` and `suspend` units only. 

Why does `get_instance` gets a parameter? So the call by the constructor/destructor does not instantiate the singleton instance (and start incrementing the expectation call count) if one was not already created for the purpose of a test case and start incrementing the count. 

In other words without what you mention `test_suspend` would look as follows:

```C
TEST_F(TestPwmOut, test_suspend)
{
    using ::testing::Mock;

    EXPECT_CALL(
        *mbed_mgmt_mock, sleep_manager_unlock_deep_sleep_internal()
    ).Times(2); // I EXPECT IT TO BE CALLED ONLY ONCE NOT TWICE, IT LOOKS WRONG
    pwm_obj->suspend();
    pwm_obj->suspend();

    // Force gMock to verify a mock object singleton before it is destructed
    // by the teardown
    Mock::VerifyAndClearExpectations(mbed_mgmt_mock);
    // There will not be a leak since the destructor releases the singleton
    Mock::AllowLeak(mbed_mgmt_mock);
}
```
Instead of what it is now:

```C
TEST_F(TestPwmOut, test_suspend)
{
    using ::testing::Mock;

    EXPECT_CALL(
        *mbed_mgmt_mock, sleep_manager_unlock_deep_sleep_internal()
    ).Times(1);
    pwm_obj->suspend();
    pwm_obj->suspend();

    // Force gMock to verify a mock object singleton before it is destructed
    // by the teardown
    Mock::VerifyAndClearExpectations(mbed_mgmt_mock);
    // There will not be a leak since the destructor releases the singleton
    Mock::AllowLeak(mbed_mgmt_mock);
}
```

If we are interested in testing the constructor then I would have another test like this:

```C
TEST_F(TestPwmOut, test_constructor)
{
    using ::testing::Mock;

    EXPECT_CALL(
        *mbed_mgmt_mock, sleep_manager_lock_deep_sleep_internal()
    ).Times(1);
    PwmOut pwm(PTC1);

    // Force gMock to verify a mock object singleton before it is destructed
    // by the teardown
    Mock::VerifyAndClearExpectations(mbed_mgmt_mock);
    // There will not be a leak since the destructor releases the singleton
    Mock::AllowLeak(mbed_mgmt_mock);
}
```